File: backend/agents/sub_agents/viral.py
from datetime import datetime
from utils.api_clients import tavily_search
from utils.parsers import Attraction
from cache.redis import get_cached, set_cached, attraction_cache_key


async def fetch_viral(city: str) -> list[Attraction]:
    key = attraction_cache_key(city, "viral")
    cached = await get_cached(key)
    if cached:
        return [Attraction(**a) for a in cached]

    year = datetime.now().year
    data = await tavily_search(
        f"viral Instagram spots hidden gems must-see {city} {year}", 8
    )

    attractions = [
        Attraction(
            id=f"viral-{i}",
            name=r.get("title", "").split(" - ")[0],
            category="viral",
            description=str(r.get("content", r.get("snippet", "")))[:250],
            url=r.get("url", ""),
        )
        for i, r in enumerate(data.get("results", []))
    ]

    await set_cached(key, [a.to_dict() for a in attractions])
    return attractions

File: src/lib/agents/sub-agents/viral.ts
import { tavilySearch } from "@/lib/utils/api-clients";
import type { Attraction } from "@/lib/utils/parsers";
import { getCached, setCached, attractionCacheKey } from "@/lib/cache/redis";

export interface ViralParams {
  city: string;
}

export async function fetchViral(params: ViralParams): Promise<Attraction[]> {
  const key = attractionCacheKey(params.city, "viral");
  const cached = await getCached<Attraction[]>(key);
  if (cached) return cached;

  const data = await tavilySearch(
    `viral Instagram spots hidden gems must-see ${params.city} ${new Date().getFullYear()}`,
    8
  );

  const attractions: Attraction[] = (data.results ?? []).map(
    (r: Record<string, unknown>, i: number) => ({
      id: `viral-${i}`,
      name: String(r.title ?? "").split(" - ")[0],
      category: "viral",
      description: String(r.content ?? r.snippet ?? "").slice(0, 250),
      url: String(r.url ?? ""),
    })
  );

  await setCached(key, attractions);
  return attractions;
}
